renku/ui/service/gateways/repository_cache.py

```diff
@@ -0,0 +1,233 @@
+# Copyright Swiss Data Science Center (SDSC). A partnership between
+# École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Repository cache interface."""
+
+import os
+import shutil
+import uuid
+from datetime import datetime
+from typing import Optional
+from urllib.parse import urlparse
+
+import portalocker
+from marshmallow import ValidationError
+
+from renku.command.clone import project_clone_command
+from renku.core import errors
+from renku.core.util.contexts import renku_project_context
+from renku.core.util.os import normalize_to_ascii
+from renku.domain_model.git import GitURL
+from renku.infrastructure.repository import Repository
+from renku.ui.service.cache import ServiceCache
+from renku.ui.service.cache.models.project import Project
+from renku.ui.service.cache.models.user import User
+from renku.ui.service.config import PROJECT_CLONE_DEPTH_DEFAULT
+from renku.ui.service.errors import IntermittentCacheError, IntermittentLockError
+from renku.ui.service.interfaces.repository_cache import IRepositoryCache
+from renku.ui.service.logger import service_log
+
+
+class LocalRepositoryCache(IRepositoryCache):
+    """Cache for project repos stored on local disk."""
+
+    def get(
+        self, cache: ServiceCache, git_url: str, branch: Optional[str], user: User, shallow: bool = True
+    ) -> Project:
+        """Get a project from cache (clone if necessary)."""
+        if git_url is None:
+            raise ValidationError("Invalid `git_url`, URL is empty", "git_url")
+
+        try:
+            project = Project.get(
```

Comment:
I am not really familiar with Walrus. But the author of the package says this in the docs about using `Models`:

> Walrus models should not be considered production-grade code and I strongly advise against anyone actually using it for anything other than experimenting or for inspiration/learning.

> My advice: just use hashes for your structured data. If you need ad-hoc queries, then use a relational database.

Should we worry about this?

See here: https://walrus.readthedocs.io/en/latest/models.html#

Reply:
Ok 2 years without issues justifies leaving this alone then.